File: sample/src/main/java/com/anggrayudi/materialpreference/sample/DonationActivity.java
package com.anggrayudi.materialpreference.sample;

import android.os.Bundle;
import android.view.LayoutInflater;
import android.view.MenuItem;
import android.view.View;
import android.view.ViewGroup;
import android.widget.Button;
import android.widget.TextView;

import com.afollestad.materialdialogs.MaterialDialog;
import com.android.billingclient.api.BillingClient;
import com.android.billingclient.api.Purchase;
import com.anggrayudi.materialpreference.sample.billing.BillingManager;
import com.anggrayudi.materialpreference.sample.billing.Constants;
import com.anggrayudi.materialpreference.sample.billing.DonationItem;

import java.util.List;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.appcompat.app.AppCompatActivity;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;

public class DonationActivity extends AppCompatActivity implements BillingManager.BillingUpdatesListener {

    private BillingManager mBillingManager;

    @Override
    protected void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_donation);
        setSupportActionBar(findViewById(R.id.toolbar));
        getSupportActionBar().setDisplayHomeAsUpEnabled(true);

        mBillingManager = new BillingManager(this, this);

        RecyclerView recyclerView = findViewById(R.id.recyclerView);
        recyclerView.setLayoutManager(new LinearLayoutManager(this));
        recyclerView.setAdapter(new DonationAdapter());
    }

    @Override
    public boolean onOptionsItemSelected(MenuItem item) {
        switch (item.getItemId()) {
            case android.R.id.home:
                onBackPressed();
                break;
        }
        return super.onOptionsItemSelected(item);
    }

    @Override
    public void onBackPressed() {
        super.onBackPressed();
        if (isFinishing()) {
            mBillingManager.destroy();
        }
    }

    @Override
    public void onBillingClientSetupFinished() {
        if (mBillingManager.mBillingClientResponseCode != BillingClient.BillingResponse.OK) {
            setRequestedOrientation(getResources().getConfiguration().orientation);
            new MaterialDialog.Builder(this)
                    .cancelable(false)
                    .content("Google Play is not setup correctly")
                    .positiveText(android.R.string.ok)
                    .onPositive((dialog, which) -> finish())
                    .show();
        }
    }

    @Override
    public void onConsumeFinished(String token, int result) {
        if (result == BillingClient.BillingResponse.OK) {
            new MaterialDialog.Builder(this)
                    .title("Thank you")
                    .content("We have received your donation. All donations you gave are really helpful for us to develop this library.")
                    .positiveText(android.R.string.ok)
                    .show();
        }
    }

    @Override
    public void onPurchasesUpdated(List<Purchase> purchases) {
        for (Purchase purchase : purchases) {
            mBillingManager.consumeAsync(purchase.getPurchaseToken());
        }
    }

    private class DonationAdapter extends RecyclerView.Adapter<DonationAdapter.ViewHolder> implements
            View.OnClickListener {

        class ViewHolder extends RecyclerView.ViewHolder {
            final TextView title, price, description;
            final Button btnDonate;

            ViewHolder(@NonNull View itemView) {
                super(itemView);
                title = itemView.findViewById(R.id.itemTitle);
                price = itemView.findViewById(R.id.itemPrice);
                description = itemView.findViewById(R.id.description);
                btnDonate = itemView.findViewById(R.id.btnDonate);
                btnDonate.setOnClickListener(DonationAdapter.this);
            }
        }

        @Override
        public void onClick(View v) {
            if (mBillingManager.mBillingClientResponseCode == BillingClient.BillingResponse.OK) {
                String sku = v.getTag().toString();
                mBillingManager.initiatePurchaseFlow(sku, BillingClient.SkuType.INAPP);
            }
        }

        @NonNull
        @Override
        public ViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
            return new ViewHolder(LayoutInflater.from(parent.getContext())
                    .inflate(R.layout.item_donation, parent, false));
        }

        @Override
        public void onBindViewHolder(@NonNull ViewHolder holder, int position) {
            DonationItem item = Constants.DONATION_ITEMS[position];
            holder.title.setText(item.title);
            holder.price.setText(item.price);
            holder.description.setText(item.description);
            holder.btnDonate.setTag(item.sku);
        }

        @Override
        public int getItemCount() {
            return Constants.DONATION_ITEMS.length;
        }
    }
}

File: sample/src/main/java/com/anggrayudi/materialpreference/sample/DonationActivity.kt
package com.anggrayudi.materialpreference.sample

import android.content.Intent
import android.net.Uri
import android.os.Bundle
import android.view.*
import android.widget.Button
import android.widget.TextView
import androidx.appcompat.app.AppCompatActivity
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.recyclerview.widget.RecyclerView
import com.afollestad.materialdialogs.MaterialDialog
import com.android.billingclient.api.BillingClient
import com.android.billingclient.api.Purchase
import com.anggrayudi.materialpreference.sample.billing.BillingManager
import com.anggrayudi.materialpreference.sample.billing.DonationItem

class DonationActivity : AppCompatActivity(), BillingManager.BillingUpdatesListener {

    private var mBillingManager: BillingManager? = null

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_donation)
        setSupportActionBar(findViewById(R.id.toolbar))
        supportActionBar!!.setDisplayHomeAsUpEnabled(true)

        mBillingManager = BillingManager(this, this)

        val recyclerView = findViewById<RecyclerView>(R.id.recyclerView)
        recyclerView.layoutManager = LinearLayoutManager(this)
        recyclerView.adapter = DonationAdapter()
    }

    override fun onCreateOptionsMenu(menu: Menu): Boolean {
        menuInflater.inflate(R.menu.donation, menu)
        menu.findItem(R.id.action_donate_paypal).intent = Intent(Intent.ACTION_VIEW,
                Uri.parse("https://www.paypal.com/cgi-bin/webscr?cmd=_s-xclick&hosted_button_id=TGPGSY66LKUMN&source=url"))
        return super.onCreateOptionsMenu(menu)
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        when (item.itemId) {
            android.R.id.home -> onBackPressed()
        }
        return super.onOptionsItemSelected(item)
    }

    override fun onBackPressed() {
        super.onBackPressed()
        if (isFinishing) {
            mBillingManager?.destroy()
        }
    }

    override fun onBillingClientSetupFinished() {
        if (mBillingManager!!.billingClientResponseCode != BillingClient.BillingResponse.OK) {
            requestedOrientation = resources.configuration.orientation
            MaterialDialog(this)
                    .cancelable(false)
                    .message(text = "Google Play is not setup correctly")
                    .positiveButton(android.R.string.ok) {finish()}
                    .show()
        }
    }

    override fun onConsumeFinished(token: String, result: Int) {
        if (result == BillingClient.BillingResponse.OK) {
            MaterialDialog(this)
                    .title(text = "Thank you")
                    .message(text = "We have received your donation. All donations you gave are really helpful for us to develop this library.")
                    .positiveButton(android.R.string.ok)
                    .show()
        }
    }

    override fun onPurchasesUpdated(purchases: List<Purchase>) {
        purchases.forEach { mBillingManager!!.consumeAsync(it.purchaseToken) }
    }

    private inner class DonationAdapter : RecyclerView.Adapter<DonationAdapter.ViewHolder>(), View.OnClickListener {

        internal inner class ViewHolder(itemView: View) : RecyclerView.ViewHolder(itemView) {
            val title: TextView = itemView.findViewById(R.id.itemTitle)
            val price: TextView = itemView.findViewById(R.id.itemPrice)
            val btnDonate: Button = itemView.findViewById(R.id.btnDonate)
            init {
                btnDonate.setOnClickListener(this@DonationAdapter)
            }
        }

        override fun onClick(v: View) {
            if (mBillingManager!!.billingClientResponseCode == BillingClient.BillingResponse.OK) {
                val sku = v.tag.toString()
                mBillingManager!!.initiatePurchaseFlow(sku, BillingClient.SkuType.INAPP)
            }
        }

        override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): ViewHolder {
            return ViewHolder(LayoutInflater.from(parent.context)
                    .inflate(R.layout.item_donation, parent, false))
        }

        override fun onBindViewHolder(holder: ViewHolder, position: Int) {
            val item = DonationItem.items[position]
            holder.title.text = item.title
            holder.price.text = item.price
            holder.btnDonate.tag = item.sku
        }

        override fun getItemCount() = DonationItem.items.size
    }
}
